# Puzzle
https://thefiddler.substack.com/p/can-you-spell-like-a-queen-bee

### This Week’s Fiddler
In Spelling Bee, a word game from The New York Times, you must create words using seven letters arranged in a honeycomb grid. Letters can be used more than once per word, but each word must be at least four letters long and include the central letter. You may recall a previous puzzle (RIP, FiveThirtyEight) based on Spelling Bee.

Each day, your goal is to score as many points as possible. Four-letter words are worth 1 point, longer words are worth the number of letters they contain, and “pangrams” (words containing every letter) provide 7 bonus points. But these specific details don’t matter for this week’s puzzle.

If you find all the words in a given day and therefore accrue the maximum number of points, you earn the “Queen Bee” ranking. Meanwhile, accruing smaller point totals earns you other rankings. In particular, the point cutoff for “Genius” is 70 percent of the maximum number of points, rounded to the nearest whole number. While the cutoff for “Genius” is clearly displayed in the puzzle, the “Queen Bee” point total is not readily shown.

Of course, this maximum total can be approximated by dividing the “Genius” cutoff by 0.7. Even so, the total may be ambiguous, since multiple “Queen Bee” values can result in the same “Genius” cutoff.

Suppose a given round of Spelling Bee has some very large, randomly chosen point total. What is the probability that this total can be precisely determined (i.e., without any ambiguity) from its point cutoff for “Genius”?

### This Week’s Extra Credit
Let’s put the rank of “Genius” aside. Here are some other ranks you can attain in Spelling Bee:

Amazing (if you get 50 percent of the maximum, rounded to the nearest whole number)

Great (40 percent)

Nice (25 percent)

Solid (15 percent)

Good (8 percent)

Moving Up (5 percent)

Good Start (2 percent)

Suppose a given round of Spelling Bee has some very large, randomly chosen point total. What is the probability that this total can be precisely determined from these cutoffs (i.e., from “Good Start” through “Amazing,” inclusive)?

### Clarification

Rounding convention is that 0.5 is rounded up.

# Solution

While the fiddler could be done by hand (and I think it is 6 ambigous to 4 unambiguous), the EC likely needs code for one to be sure, so I am just going the code route.

In [1]:
def round_half_to_pinf(x):
    return int(x + 0.5)

def cutoff_from_fullscore(pct, fullscore):
    cutoff_unrounded = fullscore * pct / 100.0
    return round_half_to_pinf(cutoff_unrounded)

def get_cutoffs(pcts, fullscore):
    cutoffs = [cutoff_from_fullscore(pct, fullscore) for pct in pcts]
    return tuple(cutoffs)

def get_range_maps(pcts, min_score=100, max_score=200):
    full_to_cutoffs = {}
    cutoffs_to_full = {}
    for full in range(min_score, max_score):
        cutoffs = get_cutoffs(pcts, full)
        full_to_cutoffs[full] = cutoffs
        if cutoffs not in cutoffs_to_full:
            cutoffs_to_full[cutoffs] = []
        cutoffs_to_full[cutoffs].append(full)
    return full_to_cutoffs, cutoffs_to_full

import math
def one_recurrence_interval(p):
    i = 1
    while i * 100 % p != 0:
        i += 1
    return int((i * 100) / p)

def analyze_cutoffs(pcts, scale=10):
    recurrence_intervals = [one_recurrence_interval(p) for p in pcts]
    recurrence_interval = math.lcm(*recurrence_intervals)
    print(f"Combined Recurrence interval: {recurrence_interval}")
    min_score = recurrence_interval * 1 * scale
    min_counting_threshold = recurrence_interval * 2 * scale
    max_counting_threshold = recurrence_interval * 3 * scale
    max_score = recurrence_interval * 4 * scale
    full_to_cutoffs, cutoffs_to_full = get_range_maps(pcts, min_score, max_score)
    num_full_scores = 0
    num_unique_cutoffs = 0
    ambiguous_cutoffs = 0
    ambiguous_full_scores = 0
    unambiguous_cutoffs = 0
    unambiguous_full_scores = 0
    for cutoffs, full_scores in cutoffs_to_full.items():
        if min_counting_threshold <= full_scores[0] < max_counting_threshold:
            num_full_scores += len(full_scores)
            num_unique_cutoffs += 1
            len_full_scores = len(full_scores)
            #print(f"Cutoff: {cutoff}, Full scores: {full_scores}")
            if len_full_scores > 1:
                ambiguous_cutoffs += 1
                ambiguous_full_scores += len_full_scores
            else:
                unambiguous_cutoffs += 1
                unambiguous_full_scores += len_full_scores
    print(f"Full scores: Total: {num_full_scores}, Ambiguous: {ambiguous_full_scores}, Unambiguous: {unambiguous_full_scores}, Unambiguous probability: {1.0 * unambiguous_full_scores / num_full_scores}")
    print(f"    Cutoffs: Total: {num_unique_cutoffs}, Ambiguous: {ambiguous_cutoffs}, Unambiguous: {unambiguous_cutoffs}, Unambiguous probability: {1.0 * unambiguous_cutoffs / num_unique_cutoffs}")

In [2]:
# Fiddler
for scale in [1, 2, 5, 10, 20, 50, 100]:
    analyze_cutoffs([70], scale=scale)

Combined Recurrence interval: 10
Full scores: Total: 10, Ambiguous: 6, Unambiguous: 4, Unambiguous probability: 0.4
    Cutoffs: Total: 7, Ambiguous: 3, Unambiguous: 4, Unambiguous probability: 0.5714285714285714
Combined Recurrence interval: 10
Full scores: Total: 20, Ambiguous: 12, Unambiguous: 8, Unambiguous probability: 0.4
    Cutoffs: Total: 14, Ambiguous: 6, Unambiguous: 8, Unambiguous probability: 0.5714285714285714
Combined Recurrence interval: 10
Full scores: Total: 50, Ambiguous: 30, Unambiguous: 20, Unambiguous probability: 0.4
    Cutoffs: Total: 35, Ambiguous: 15, Unambiguous: 20, Unambiguous probability: 0.5714285714285714
Combined Recurrence interval: 10
Full scores: Total: 100, Ambiguous: 60, Unambiguous: 40, Unambiguous probability: 0.4
    Cutoffs: Total: 70, Ambiguous: 30, Unambiguous: 40, Unambiguous probability: 0.5714285714285714
Combined Recurrence interval: 10
Full scores: Total: 200, Ambiguous: 120, Unambiguous: 80, Unambiguous probability: 0.4
    Cutoffs: To

In [3]:
# Extra credit
for scale in [1, 2, 5, 10, 20, 50, 100]:
    analyze_cutoffs([2,5,8,15, 25, 40, 50], scale=scale)

Combined Recurrence interval: 100
Full scores: Total: 100, Ambiguous: 30, Unambiguous: 70, Unambiguous probability: 0.7
    Cutoffs: Total: 85, Ambiguous: 15, Unambiguous: 70, Unambiguous probability: 0.8235294117647058
Combined Recurrence interval: 100
Full scores: Total: 200, Ambiguous: 60, Unambiguous: 140, Unambiguous probability: 0.7
    Cutoffs: Total: 170, Ambiguous: 30, Unambiguous: 140, Unambiguous probability: 0.8235294117647058
Combined Recurrence interval: 100
Full scores: Total: 500, Ambiguous: 150, Unambiguous: 350, Unambiguous probability: 0.7
    Cutoffs: Total: 425, Ambiguous: 75, Unambiguous: 350, Unambiguous probability: 0.8235294117647058
Combined Recurrence interval: 100
Full scores: Total: 1000, Ambiguous: 300, Unambiguous: 700, Unambiguous probability: 0.7
    Cutoffs: Total: 850, Ambiguous: 150, Unambiguous: 700, Unambiguous probability: 0.8235294117647058
Combined Recurrence interval: 100
Full scores: Total: 2000, Ambiguous: 600, Unambiguous: 1400, Unambiguous 

# Conclusion

Took some effort to get the counting correct for the EC :) *Some cases at the edges were falsely marked as unambiguous when it was just that I hadn't considered the values that would make them ambiguous.*

**Fiddler Answer = 0.4**

**Extra Credit Answer = 0.7**

This is assuming that the sample space is such that the max points total is uniformly distributed. If it were that we were picking uniformly from among the legal cutoff values, the probabilities would be higher - 0.5714285714285714 and 0.8235294117647058 respectively for the Fiddler and the EC.